File: day11/main.go
package main

import "aocutil"

//generates a "aabbccdd.." sequence starting from the given index
func default_sequence(input *[]byte, start int) {
	length := len(*input) - start
	c := byte('a') - 1
	for i := 0; i < length; i++ {
		if i%2 == 0 {
			c++
		}
		(*input)[i+start] = c
	}
}

func increase(input *[]byte) {
	last := len(*input) - 1
	for i := last; i > 0; i-- {
		if (*input)[i] == 'z' {
			(*input)[i] = 'a'
		} else {
			(*input)[i]++
			break
		}
	}
}

func next_password(input string) string {
	arr := []byte(input)
	increase(&arr)

	changed := true
	for changed {
		changed = false
		//check for invalid characters
		for i := 0; i < len(arr); i++ {
			if arr[i] == 'i' || arr[i] == 'o' || arr[i] == 'l' {
				//if invalid increase char by 1 (i/o/l cannot wrap)
				arr[i]++
				//and change following sequence to aabbcc...
				i++
				default_sequence(&arr, i)
				changed = true
				break
			}
		}

		//has no invalid characters anymore at this point
		//check for increasing straight pattern "abc" "bcd" etc
		hasIncreasingPattern := false
		increasingPatternCount := 1
		for i := 0; i < len(arr)-1; i++ {
			if arr[i]+1 == arr[i+1] {
				increasingPatternCount++
			} else {
				if increasingPatternCount >= 3 {
					hasIncreasingPattern = true
					break
				}
				increasingPatternCount = 1
			}
		}
		//if the whole password is a sequence it will never enter the else branch
		//check again here
		if increasingPatternCount >= 3 {
			hasIncreasingPattern = true
		}

		differentPairsCount := 0
		previous := byte(0)
		for i := 0; i < len(arr)-1; i++ {
			if previous != arr[i] && arr[i] == arr[i+1] {
				differentPairsCount++
				previous = arr[i]
				i++
			}
		}

		if !hasIncreasingPattern || differentPairsCount < 2 {
			increase(&arr)
			changed = true
		}
	}

	return string(arr)
}

func main() {
	input := "hxbxwxba"
	part1 := next_password(input)
	part2 := next_password(part1)
	aocutil.AOCFinish(part1, part2)
}

File: day11/main.c
#include <aoc/common.h>
#include <stdio.h>
#include <aoc/aoc.h>
#include <aoc/mem.h>

static const char input[] = "hxbxwxba";
static const i32 input_len = (i32)AOC_ARRAY_SIZE(input) - 1;

static inline void default_sequence(char pwd[input_len], const i32 start) {
  char c = 'a' - 1;
  for (i32 i = start; i < input_len; ++i) {
    c += (i & 1);
    pwd[i] = c;
  }
}

static inline void increase(char pwd[input_len]) {
  for (i32 i = input_len - 1; i > 0; --i) {
    if (pwd[i] == 'z') {
      pwd[i] = 'a';
    } else {
      pwd[i]++;
      break;
    }
  }
}

static inline bool invalid_character_check(char pwd[const input_len]) {
  for (i32 i = 0; i < input_len; ++i) {
    if (pwd[i] == 'i' || pwd[i] == 'o' || pwd[i] == 'l') {
      pwd[i]++;
      i++;
      default_sequence(pwd, i);
      return false;
    }
  }
  return true;
}

static inline bool increasing_pattern_check(char pwd[const input_len]) {
  i8 increasingPatternCount = 1;
  for (i32 i = 0; i < input_len - 1; ++i) {
    if (pwd[i] + 1 == pwd[i + 1]) {
      increasingPatternCount++;
    } else {
      if (increasingPatternCount >= 3)
        break;
      increasingPatternCount = 1;
    }
  }
  return increasingPatternCount >= 3;
}

static inline bool different_pairs_check(char pwd[const input_len]) {
  i8 differentPairsCount = 0;
  char previous = '0';
  for (i32 i = 0; i < input_len - 1; ++i) {
    if (previous != pwd[i] && pwd[i] == pwd[i + 1]) {
      differentPairsCount++;
      previous = pwd[i++];
    }
  }
  return differentPairsCount >= 2;
}

static void next_password(const char pwd[const input_len],
                          char output[const input_len]) {
  aoc_mem_copy(output, pwd, input_len);
  for (;;) {
    increase(output);
    if (invalid_character_check(output) && increasing_pattern_check(output) &&
        different_pairs_check(output))
      break;
  }
}

int main(void) {
  char part1[input_len];
  next_password(input, part1);
  char part2[input_len];
  next_password(part1, part2);
  printf("%.*s\n%.*s\n", input_len, part1, input_len, part2);
}
